File: frontend/src/api/timeline.ts
// Mirrors backend/app/schemas/timeline.py (serialized as media/{debate_id}/timeline.json).

export type SpeakerRole = 'moderator' | 'debater' | 'judge';
export type Mascot = 'orb' | 'bolt' | 'cube';

export interface TimelineWord {
  w: string;
  s: number; // ms relative to the segment audio
  e: number;
}

export interface TimelineSpeaker {
  id: string;
  name: string;
  role: SpeakerRole;
  model: string;
  color: string;
  mascot: Mascot;
  avatar_url: string | null;
  voice_id: string | null;
}

export interface TimelineSegment {
  seq_index: number;
  speaker_id: string;
  speaker_name: string;
  round_id: string;
  turn_type: string;
  start_ms: number; // absolute on the full track
  end_ms: number;
  audio: string;
  text: string;
  words: TimelineWord[];
  levels: number[];
  levels_hz: number;
  note: string | null;
}

export interface TimelineHighlight {
  index: number;
  title: string;
  hook: string;
  start_ms: number;
  end_ms: number;
  seq_indexes: number[];
}

export interface TimelineVerdict {
  seq_index: number;
  winner_id: string | null;
  winner_name: string | null;
}

export interface Timeline {
  version: 1;
  debate_id: string;
  title: string;
  topic: string;
  language: string;
  language_code: string;
  created_at: string;
  provider: string;
  model_id: string;
  speakers: TimelineSpeaker[];
  segments: TimelineSegment[];
  gap_ms: number;
  total_ms: number;
  full_audio_wav: string;
  full_audio_mp3: string;
  verdict: TimelineVerdict | null;
  highlights: TimelineHighlight[];
  stats: { chars: number; tts_ms: number; estimated_usd: number | null; cached_turns: number };
}

File: backend/app/schemas/timeline.py
"""Timeline: the contract between the audio build (worker), the API and the browser-side
video renderer. Serialized to `media/{debate_id}/timeline.json`."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

SpeakerRole = Literal["moderator", "debater", "judge"]
Mascot = Literal["orb", "bolt", "cube"]


class TimelineWord(BaseModel):
    w: str
    s: int  # ms relative to the segment's own audio
    e: int


class TimelineSpeaker(BaseModel):
    id: str  # "participant_{i}" | "judge"
    name: str
    role: SpeakerRole
    model: str
    color: str
    mascot: Mascot
    avatar_url: str | None = None
    voice_id: str | None = None


class TimelineSegment(BaseModel):
    seq_index: int
    speaker_id: str
    speaker_name: str
    round_id: str
    turn_type: str
    start_ms: int  # absolute on the full track
    end_ms: int
    audio: str  # relative path, e.g. "turns/003.wav"
    text: str  # spoken text without audio tags
    words: list[TimelineWord] = Field(default_factory=list)
    levels: list[float] = Field(default_factory=list)  # RMS envelope 0..1
    levels_hz: int = 20
    note: str | None = None


class TimelineHighlight(BaseModel):
    index: int
    title: str
    hook: str
    start_ms: int
    end_ms: int
    seq_indexes: list[int] = Field(default_factory=list)


class TimelineVerdict(BaseModel):
    seq_index: int
    winner_id: str | None = None
    winner_name: str | None = None


class TimelineStats(BaseModel):
    chars: int = 0
    tts_ms: int = 0
    estimated_usd: float | None = None
    cached_turns: int = 0


class Timeline(BaseModel):
    version: Literal[1] = 1
    debate_id: str
    title: str
    topic: str
    language: str
    language_code: str
    created_at: datetime
    provider: str
    model_id: str
    speakers: list[TimelineSpeaker]
    segments: list[TimelineSegment]
    gap_ms: int
    total_ms: int
    full_audio_wav: str = "full.wav"
    full_audio_mp3: str = "full.mp3"
    verdict: TimelineVerdict | None = None
    highlights: list[TimelineHighlight] = Field(default_factory=list)
    stats: TimelineStats = Field(default_factory=TimelineStats)
